op-challenger/game/keccak/matrix/keccak.go

```diff
@@ -0,0 +1,59 @@
+package matrix
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// StateMatrix implements a stateful keccak sponge with the ability to create state commitments after each permutation
+type StateMatrix struct {
+	s *state
+}
+
+// LeafSize is the size in bytes required for leaf data.
+const LeafSize = 136
+
+var uint256Size = 32
+
+// NewStateMatrix creates a new state matrix initialized with the initial, zero keccak block.
+func NewStateMatrix() *StateMatrix {
+	return &StateMatrix{s: newLegacyKeccak256()}
+}
+
+// StateCommitment returns the state commitment for the current state matrix.
+// Additional data may be absorbed after calling this method.
+func (d *StateMatrix) StateCommitment() common.Hash {
+	buf := d.packState()
+	return crypto.Keccak256Hash(buf)
+}
+
+func (d *StateMatrix) packState() []byte {
+	buf := make([]byte, 0, len(d.s.a)*uint256Size)
+	for _, v := range d.s.a {
+		buf = append(buf, math.U256Bytes(new(big.Int).SetUint64(v))...)
+	}
+	return buf
+}
+
+// AbsorbLeaf absorbs the specified data into the keccak sponge.
+// If final is true, the data is padded to the required length, otherwise it must be exactly
+// LeafSize bytes.
+func (d *StateMatrix) AbsorbLeaf(data []byte, final bool) {
+	if !final && len(data) != LeafSize {
+		panic("sha3: Incorrect leaf data length")
+	}
```

Comment:
I'm not against making this return an error but:
a. It is a program error to supply incorrect data length. Not hard to get this right for any length input.
b. The underlying sha3 functions copied from the go libraries panic on invalid usage for other reasons already (e.g adding data after calling `Read`) and I'm keen to minimise modifications to the keccak code